plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs

```diff
@@ -73,230 +151,453 @@ their margin. If either party fails to make a margin payment then the contract
+                    Validation.valueLockedBy ptx ftoLongAccount `geq` payoutsLong
+                    && Validation.valueLockedBy ptx ftoShortAccount `geq` payoutsShort
+                slotvalid   = oracleDate == ftDeliveryDate
+                                && ftDeliveryDate `Interval.before` pendingTxValidRange ptx
+            in slotvalid && payoutValid
+        _ -> False
 
 -- | Compute the required margin from the current price of the
 --   underlying asset.
-requiredMargin :: Future -> Ada -> Ada
-requiredMargin Future{futureUnits=units, futureUnitPrice=unitPrice, futureMarginPenalty=pnlty} spotPrice =
+{-# INLINABLE requiredMargin #-}
+requiredMargin :: Future -> Value -> Value
+requiredMargin Future{ftUnits, ftUnitPrice, ftMarginPenalty} spotPrice =
     let
-        delta  = (Ada.lovelaceOf units) * (spotPrice - unitPrice)
+        delta  = scale ftUnits (spotPrice - ftUnitPrice)
     in
-        pnlty + delta
-
-{-# INLINABLE mkValidator #-}
-mkValidator :: Future -> FutureData -> FutureRedeemer -> PendingTx -> Bool
-mkValidator ft@Future{..} FutureData{..} r p@PendingTx{pendingTxOutputs=outs, pendingTxValidRange=range} =
+        ftMarginPenalty + delta
+
+{-# INLINABLE initialMargin #-}
+initialMargin :: Future -> Value
+initialMargin ft@Future{ftUnitPrice, ftMarginPenalty} =
+    ftMarginPenalty + ftUnitPrice
+
+{-# INLINABLE initialState #-}
+-- | The initial state of the 'Future' contract
+initialState :: Future -> FutureState
+initialState ft =
+    let im = initialMargin ft in
+    Running (MarginAccounts{ftsShortMargin=im, ftsLongMargin=im})
+
+futureAddress :: Future -> FutureOwners -> Address
+futureAddress ft fo = Ledger.scriptAddress (validatorScript ft fo)
+
+-- | The data needed to initialise the futures contract.
+data FutureSetup =
+    FutureSetup
+        { shortPK :: PubKey
+        -- ^ Initial owner of the short token
+        , longPK :: PubKey
+        -- ^ Initial owner of the long token
+        , contractStart :: Slot
+        -- ^ Start of the futures contract itself. By this time the setup code 
+        --   has to be finished, otherwise the contract is void.
+        } deriving (Haskell.Show)
+
+-- | Initialise the contract by
+--   * Generating the tokens for long and short
+--   * Setting up an escrow contract for the initial margins
+--   * Paying the initial margin for the given role
+initialiseFuture
+    :: ( HasEndpoint "initialise-future" (FutureSetup, Role) s
+       , HasBlockchainActions s
+       , AsFutureError e
+       )
+    => Future
+    -> Contract s e FutureOwners
+initialiseFuture future = do
+    (s, ownRole) <- endpoint @"initialise-future" @(FutureSetup, Role)
+    ftos <- setupTokens
+    let escr = escrowParams future ftos s
+        escrowPayment = Escrow.payRedeemRefund escr (initialMargin future <> tokenFor Long ftos <> tokenFor Short ftos)
+    e <- withContractError (review _EscrowFailed) escrowPayment
+    either (throwing _EscrowRefunded) (\_ -> pure ftos) e
```

Comment:
I think these all could do with inline comments explaining the stages, since both:
- The mechanism of the contract is quite complicated
- These functions pull together quite a lot of pieces.